src/vector.cpp

```diff
@@ -8,122 +8,97 @@ extern "C" {
-            vecbasic_concentrate_vec(out, VECTOR_ELT(dots, i));
-        } else {
-            Rf_error("The %dth object's type is error'", i + 1);
-        }
+// [[Rcpp::export(".vecbasic_subset")]]
+SEXP sexp_vecbasic_subset(SEXP ext, SEXP idx) {
+    if (TYPEOF(idx) != INTSXP)
+        Rf_error("Internal? Index must be integer");
+    
+    CVecBasic* inv  = (CVecBasic*) R_ExternalPtrAddr(ext);
+    SEXP       out  = PROTECT(sexp_vecbasic());
+    CVecBasic* outv = (CVecBasic*) R_ExternalPtrAddr(out);
+    int*       ids  = INTEGER(idx);
+    
+    // Used as a temporarily value in the loop
+    SEXP a = PROTECT(sexp_basic());
+    basic_struct* val = (basic_struct*) R_ExternalPtrAddr(a);
+    for (int i = 0; i < Rf_length(idx); i++) {
+        vecbasic_get(inv, ids[i] - 1, val);
+        vecbasic_push_back(outv, val);
     }
-    UNPROTECT(1);
+    
+    UNPROTECT(2);
     return out;
 }
 
-
 // [[Rcpp::export(".vecbasic_get")]]
-SEXP vecbasic_get(SEXP ext, SEXP n) {
-    // It can't handle out_of_range for now.
-    SEXP   out = PROTECT(sexp_basic());
-    sexp_vecbasic_get(ext, n, out);
+SEXP sexp_vecbasic_get(SEXP ext, SEXP n) {
+    SEXP          out  = PROTECT(sexp_basic());
+    basic_struct* outv = (basic_struct*) R_ExternalPtrAddr(out);
+    
+    // Currently vecbasic_get is implemented with vecbasic_subset, then extract the value
+    if (Rf_length(n) == 1) {
+        SEXP res = PROTECT(sexp_vecbasic_subset(ext, n));
+        vecbasic_get((CVecBasic*) R_ExternalPtrAddr(res), 0, outv);
+        UNPROTECT(1);
+    }
+    
+    if (Rf_length(n) > 1)
+        Rf_error("Attempt to select more than one element in vectorIndex");
+    
+    if (Rf_length(n) == 0)
+        Rf_error("Attempt to select less than one element in vectorIndex");
+    
     UNPROTECT(1);
     return out;
 }
 
-// [[Rcpp::export(".vecbasic_subset")]]
-SEXP vecbasic_subset(SEXP ext, SEXP dots) {
-    // It can't handle out_of_range for now.
-    SEXP   out = PROTECT(sexp_vecbasic());
-    SEXP   val = PROTECT(sexp_basic());
-    int*   lt = INTEGER(dots);
-    int    len = Rf_length(dots);
-    for (int i = 0; i < len; ++i) {
-        if (lt[i] == 0) continue;
-        sexp_vecbasic_get(ext, Rf_ScalarInteger(lt[i]), val);
-        sexp_vecbasic_push_back(out, val);
+
+static inline
+void _vecbasic_append_sexp(CVecBasic* self, SEXP ext) {
+    if ((TYPEOF(ext) == EXTPTRSXP) &&
+        R_compute_identical(R_ExternalPtrTag(ext), Rf_mkString("basic_struct*"), 15)) {
+        
+        vecbasic_push_back(self, (basic_struct*) R_ExternalPtrAddr(ext));
+        return;
     }
-    UNPROTECT(2);
-    return out;
+    if ((TYPEOF(ext) == EXTPTRSXP) &&
+        R_compute_identical(R_ExternalPtrTag(ext), Rf_mkString("CVecBasic*"), 15)) {
+        
+        CVecBasic* toappend = (CVecBasic*) R_ExternalPtrAddr(ext);
+        
+        basic_struct* val;
+        basic_new_stack(val);
```

Comment:
should be basic_struct* val = basic_new_heap(); ?

Reply:
I thought it should be okay to use basic_new_stack when we do not return it, and it may have performance advantage over basic_new_heap. See the usage at https://github.com/symengine/symengine/blob/master/symengine/tests/cwrapper/test_cwrapper.c .
However, it produces a warning in compilation and it seems to cause segfault when using in an exported function.
I have modified it using `sexp_basic()`.